src/Workspaces/Core/Portable/Shared/Extensions/SourceTextExtensions.cs

```diff
@@ -70,35 +70,64 @@ public static TextChangeRange GetEncompassingTextChangeRange(this SourceText new
         return TextChangeRange.Collapse(ranges);
     }
 
-    public static int IndexOf(this SourceText text, string value, int startIndex, bool caseSensitive)
+    public static int IndexOf(this SourceText text, string searchString, int startIndex, bool caseSensitive)
     {
-        var length = text.Length - value.Length;
-        var normalized = caseSensitive ? value : CaseInsensitiveComparison.ToLower(value);
+        if (searchString.Length == 0)
+            return -1;
+
+        var searchStringLength = searchString.Length;
+        var textEnd = text.Length - searchStringLength;
 
-        for (var i = startIndex; i <= length; i++)
+        return caseSensitive
+            ? IndexOfCaseSensitive()
+            : IndexOfCaseInsensitive();
+
+        int IndexOfCaseSensitive()
         {
-            var match = true;
-            for (var j = 0; j < normalized.Length; j++)
+            for (var i = startIndex; i <= textEnd; i++)
             {
-                // just use indexer of source text. perf of indexer depends on actual implementation of SourceText.
-                // * all of our implementation at editor layer should provide either O(1) or O(log n).
-                //
-                // only one implementation we have that could have bad indexer perf is CompositeText with heavily modified text
-                // at compiler layer but I believe that being used in find all reference will be very rare if not none.
-                if (!Match(normalized[j], text[i + j], caseSensitive))
+                var match = true;
+                for (var j = 0; j < searchStringLength; j++)
                 {
-                    match = false;
-                    break;
+                    if (searchString[j] != text[i + j])
+                    {
+                        match = false;
+                        break;
+                    }
                 }
+
+                if (match)
+                    return i;
             }
 
-            if (match)
+            return -1;
+        }
+
+        int IndexOfCaseInsensitive()
+        {
+            // Get the first character of the value to search for, with the correct case sensitivity.  This way, we don't
+            // have to call through ToLower when we're trying to find the starting point of a word match
+            var normalizedFirstChar = CaseInsensitiveComparison.ToLower(searchString[0]);
+
+            for (var i = startIndex; i <= textEnd; i++)
             {
-                return i;
+                var match = true;
+                for (var j = 0; j < searchStringLength; j++)
+                {
+                    var matchChar = j == 0 ? normalizedFirstChar : CaseInsensitiveComparison.ToLower(searchString[j]);
```

Comment:
>CaseInsensitiveComparison.ToLower [](http://example.com/codeflow?start=67&length=33)

I'm not sure of the search characteristics, but there is a potential tradeoff that we could be calling this ToLower(char) for quite a few more chars than in searchString, right?

Reply:
yes.  but i didn't measure any problems with this.  and i view allocatoins as much worse.  most code is ascii, so we're going to fastpath all these ToLowers all the time.